File: frontend/src/lib/types/community.ts
/** Cohort benchmark domain types — R4 (IPW + Laplace DP + k-anonymity). */

export interface Benchmark {
  /** the user's annualized footprint, tonnes CO2e. */
  youTco2e: number;
  /** cohort average (similar household size + income). */
  averageTco2e: number;
  /** cohort top-20% threshold. */
  topTco2e: number;
  /** signed % the user is relative to the average (negative = below/better). */
  vsAveragePct: number;
  /** cohort size, suppressed (omitted/null) below the k-anonymity threshold. */
  cohortSize?: number | null;
  /** R4: average is selection-bias-corrected (IPW) + differentially-private. */
  privacyAdjusted?: boolean;
}

File: backend/app/schemas/community.py
"""Community schemas — mirror the frontend Benchmark type."""

from __future__ import annotations

from app.schemas.common import CamelModel


class Benchmark(CamelModel):
    """R4 cohort comparison — IPW-corrected, DP-noised, k-anonymity-gated."""

    you_tco2e: float
    average_tco2e: float
    top_tco2e: float
    vs_average_pct: float
    # suppressed (None) below the k-anonymity threshold
    cohort_size: int | None = None
    # R4: average is selection-bias-corrected (IPW) + differentially-private
    privacy_adjusted: bool = False
